projects/transformers/callbacks/sparsity.py

```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
+
+        warmup_steps = self.warmup_steps or self.prune_freq
+        self.prune_scheduler = CosineDecayPruneScheduler(
+            total_steps=args.max_steps,
+            prune_fraction=self.prune_fraction,
+            warmup_steps=warmup_steps - 1
+        )
+        self.sparse_modules = filter_modules(
+            model, include_modules=[SparseWeightsBase]
+        ).values()
+
+    def on_step_end(
+        self, args, state, control, model=None,
+        train_dataloader=None, optimizer=None, **kwargs
+    ):
+        """Prune and regrow weights every 'prune_freq' iterations."""
+
+        if state.global_step % self.prune_freq != 0:
+            self.prune_scheduler.step()
+            return
+
+        # Pre-prune sparsities.
+        param_sparsity0, mask_sparsity0 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Prune weights.
+        model.apply(rezero_weights)
+        prune_fraction = self.prune_scheduler.get_prune_fraction()
+        num_removed = global_prune_by_weight(self.sparse_modules, prune_fraction)
+        model.apply(rezero_weights)
+
+        # Post-prune sparsities.
+        param_sparsity1, mask_sparsity1 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Train for one step.
+        optimizer.zero_grad()
+        train_batch = next(iter(train_dataloader))
+        inputs_to_device(train_batch, device=args.device)
+        output = model(**train_batch)
+        output.loss.backward()
+
+        # Regrow weights
+        num_add = self.prune_scheduler.get_num_add(num_removed)
+        global_add_by_grad(self.sparse_modules, num_add)
+        self.prune_scheduler.step()
+
+        # Post-grow sparsities.
+        param_sparsity2, mask_sparsity2 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Log pruning stats.
+        actual_pruned = param_sparsity1 - param_sparsity0
+        actual_pruned_on_params = actual_pruned / mask_sparsity0
+
+        logging.info(f"RigLCallback:")
+        logging.info(f"Target: remove {prune_fraction} frac of on params")
+        logging.info(f"Actual: removed {actual_pruned_on_params} fraction of on params")
+
+        # For now, the logs are very robust to ensure pruning occurs as expected.
+        logs = dict({
+            "rigl/target_pruned_on_params": prune_fraction,
+            "rigl/actual_pruned_on_params": actual_pruned / mask_sparsity0,
+            "rigl/target_pruned_total_params": prune_fraction * mask_sparsity0,
+            "rigl/actual_pruned_total_params": actual_pruned,
+            "rigl/pre_prune_param_sparsity": param_sparsity0,
+            "rigl/pre_prune_mask_sparsity": mask_sparsity0,
+            "rigl/post_prune_param_sparsity": param_sparsity1,
+            "rigl/post_prune_mask_sparsity": mask_sparsity1,
+            "rigl/pre_grow_param_sparsity": param_sparsity2,
+            "rigl/post_grow_mask_sparsity": mask_sparsity2,
+        })
+
+        wandb.log(logs, step=state.global_step)
```

Comment:
Will this fail if I am not using wandb, or this will just be ignored?

Reply:
Good catch. It will fail. I'll fix it.